Simplify and split config.json into defaultPreferences and buildConfig

yuya-oc opened the pull request:
Before submitting, please confirm you've
 - [x] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
 - [x] completed [Mattermost Contributor Agreement](http://www.mattermost.org/mattermost-contributor-agreement/)
 - [x] executed `npm run lint:js` for proper code formatting

Please provide the following information:

**Summary**
Simplify and split config.json into defaultPreferences and buildConfig

buildConfig is never overrode by user preferences. And this PR proposes several things.

- Write config in JS
It allows to write comments.
- Obsolete `override.json`
Admins need to know its format without any comments. In this PR, what they need to do is just editing existing config files.
- Multiple default servers
They are listed on the tab bar prior to user-defined servers. And they are not listed on the settings page.

Side effect: Cache purging would not work by merging this. It would be fixed after merging `AppStateManager` of #582

**Issue link**
#618 

**Test Cases**

**Additional Notes**
https://circleci.com/gh/yuya-oc/desktop/441#artifacts

jasonblais commented:
Looks great. @yuya-oc Just one question on multiple servers, what do you mean by `They are listed on the tab bar prior to user-defined servers. And they are not listed on the settings page.`

@dmeza let us know if you have any feedback on this PR. It's a follow-up for https://github.com/mattermost/desktop/pull/586

yuya-oc commented:
@jasonblais This is an example. https://circleci.com/gh/yuya-oc/desktop/449#artifacts

```js
// buildConfig.js
defaultTeams: [
  {
    name: 'example',
    url: 'https://example.com'
  },
  {
    name: 'google',
    url: 'https://www.google.com'
  }
],
```

You would see two tabs at least. And they do not appear on the settings page.

jasonblais commented:
Ah right, makes sense. Thanks @yuya-oc, looks good to me.

dmeza commented:
@yuya-oc @jasonblais looks good to me. It allows for the same functionality, thank you.

jasonblais commented:
@yuya-oc This one is ready to merge

yuya-oc commented:
To avoid confusing, I'll wait the result of #600.

jasonblais commented:
@yuya-oc Once #600 is merged, we can also merge this. It'll just need `EnableServerManagement` setting for the admin config file, so that the end user cannot modify this setting.

jasonblais commented:
(As part of this PR, we'll also restrict users from modifying servers via config.json when `EnableServerManagement` is set to `false`.

yuya-oc commented:
Updated: https://circleci.com/gh/yuya-oc/desktop/486#artifacts

`enableServerManagement: false`: https://circleci.com/gh/yuya-oc/desktop/487#artifacts

Now you can't modify servers by editing config.json. And I added a dialog to show error when the app is unusable.

`When "enableServerManagement: false" is specified in buildConfig.js,\n"defaultTeams" must have one team at least.`

<img width="329" alt="buildconfig_error" src="https://user-images.githubusercontent.com/1412057/32496438-8fe7f7f4-c40c-11e7-8dfc-e5826b085132.png">

jasonblais commented:
@yuya-oc I get the attached dialog when opening the app

![image](https://user-images.githubusercontent.com/13119842/32499432-8ca381f6-c3a0-11e7-9e1b-fa917fbe8a8f.png)

```
Application: Mattermost 3.7.1
Platform: Windows_NT 10.0.14393 x64
SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at new PermissionManager (C:\Users\Admin\AppData\Local\mattermost\app-3.7.1\resources\app.asar\main_bundle.js:6559:31)
    at permissionRequestHandler (C:\Users\Admin\AppData\Local\mattermost\app-3.7.1\resources\app.asar\main_bundle.js:6634:29)
    at App.app.on (C:\Users\Admin\AppData\Local\mattermost\app-3.7.1\resources\app.asar\main_bundle.js:8168:54)
    at emitTwo (events.js:111:20)
    at App.emit (events.js:194:7)
```

jasonblais commented:
Also, I think there's an extra space in front of `internal` that I hadn't noticed before

![image](https://user-images.githubusercontent.com/13119842/32499502-bc7dfbcc-c3a0-11e7-8359-2162a820c038.png)

jasonblais commented:
My `permissions.json` is empty, not sure if that's related.

yuya-oc commented:
JSON error: It's weird. It means `AppData\Roadming\Mattermost\permission.json` is unexpectedly broken or a blank file. Probably the app should ignore the error because the user can allow permissions later again even if the file is broken. For now, please remove `AppData\Roadming\Mattermost\permission.json`.

I had thought the white space is intended. Should we remove it?

jasonblais commented:
I removed `permission.json`, works fine now 👍 

For the dialog, does it appear for the end user when they launch the app, `enableServerManagement` is false and `DefaultTeams` is empty?

Please also help remove the white space in front of `Internal`.

jasonblais commented:
Should I create a ticket for the app to ignore the error related to `permissions.json`?

yuya-oc commented:
Yes, so actual wrong config would be below.

```js
{
  defaultTeams: [],
  helpLink: 'https://docs.mattermost.com/help/apps/desktop-guide.html',
  enableServerManagement: false
}
```

I feel Ticket is not needed. It's easy to fix, so may I fix it on `master` apart from this PR?

jasonblais commented:
> I feel Ticket is not needed. It's easy to fix, so may I fix it on master apart from this PR?

Yes, totally fine 👍 

For the dialog, is there any way to inform the admin they made a mistake? Before they distribute it to their users.

yuya-oc commented:
We can add a script to confirm buildConfig before build distributions.

jasonblais commented:
Is that something that would be a lot of work? Could it be done as part of https://github.com/mattermost/desktop/issues/483?

That way the end user would never be affected

yuya-oc commented:
Not hard. Just writing a simple script outside of the app code.

jasonblais commented:
Cool, let's do that. So we'd have
 1. script that confirms `buildConfig` before distributing the build to end users
 2. Remove whitespace in front of `Internal`
 3. Ignore error on `permissions.json`

Final question on 1: when would the script be ran (is it automatic or would the admin run it)

yuya-oc commented:
Both applicable. Probably automatic is effective to avoid mistake. For example, at the first step of `npm run package:*`.

jasonblais commented:
@yuya-oc Sounds great, agree. Would the admin run `npm run package:*` each time before distributing the app?

yuya-oc commented:
Yeah, the commands generate .zip, .exe, ... packages. So the admin would always run them before distributing.

jasonblais commented:
Perfect 👍

yuya-oc commented:
Updated: https://circleci.com/gh/yuya-oc/desktop/489#artifacts
And https://circleci.com/gh/yuya-oc/desktop/490 should fail before testing.

I fixed 2 and 3 on master branch. 30da348e6e596af73613a77a683b195ab24db757 and 4137df4098783ace6665951de3a1e1a3c0a9d3f0

yuya-oc commented:
@jasonblais Please reload this issue before open the link.

jasonblais commented on src/main.js:
```diff
@@ -367,6 +368,14 @@ app.on('ready', () => {
+    dialog.showMessageBox({
+      type: 'error',
+      title: 'Mattermost',
+      message: 'When "enableServerManagement: false" is specified in buildConfig.js,\n"defaultTeams" must have one team at least.'
```
Do you think we need the UI dialog now that we automatically message the admin? Or does it still build the app files in which case there's a chance the app will be distributed to others?

yuya-oc replied:
Surely we added the automatic checker to the build script. However by reading the contents of `package.json`, admin can distribute the app with wrong buildConfig. I think there is no such admin, though.

So I'm okay even if the UI dialog is removed. I could not determine whether the dialog is still needed.

jasonblais replied:
Wondering if `npm run package:*` fail if the admin is distributing the app with an incorrect buildConfig? Thoughts? (Not sure if that's even feasible)

yuya-oc replied:
Yes. When using wrong buildConfig, `npm run package:*` fails.

Actually, `npm run package:*` consists from three steps.

1. Check buildConfig. (this PR)
2. Compile JavaScript with webpack.
3. Create distributable packages with electron-builder.

The step 3 is written at `package.json`, so the admin can create packages without error by manually executing only step 3. However it's intentional. So I think it's okay even if we ignore such case.

jasonblais replied:
Sounds great. If people hit issues, we can always add the dialog later, but sounds like this should be okay for the majority of the admins.

We can remove the dialog 👍

jasonblais commented on scripts/check_build_config.js:
```diff
@@ -0,0 +1,16 @@
+  if (config.enableServerManagement === false && config.defaultTeams && config.defaultTeams.length === 0) {
+    return {
+      result: false,
+      message: `When "enableServerManagement: false" is specified in buildConfig.js, "defaultTeams" must have one team at least.\n${JSON.stringify(config, null, 2)}`
```
Propose
```
Specify at least one server for "defaultTeams" in buildConfig.js when "enableServerManagement is set to false.\n${JSON.stringify(config, null, 2)}
```

Would you have an example of what `${JSON.stringify(config, null, 2)}` will output in the console?

yuya-oc replied:
It would be a formatted text of buildConfig:

```
{
  defaultTeams: [],
  helpLink: 'https://docs.mattermost.com/help/apps/desktop-guide.html',
  enableServerManagement: false
}
```

yuya-oc commented:
@jasonblais Updated. https://circleci.com/gh/yuya-oc/desktop/492 should not have any artifacts because `npm run package:*` fails due to wrong buildConfig.

jasonblais commented:
Great, thanks @yuya-oc! Looks good to me, should be good to merge 👍 

> https://circleci.com/gh/yuya-oc/desktop/492 should not have any artifacts because npm run package:* fails due to wrong buildConfig

Oh right, that makes sense.. :)

yuya-oc commented:
Thanks @jasonblais and @dmeza for many feedback!
